test/Elements.Tests/RayTests.cs

```diff
@@ -1,126 +1,162 @@
+        {
+            var a = new Vertex(new Vector3(-0.5, -0.5, 1.0));
+            var b = new Vertex(new Vector3(0.5, -0.5, 1.0));
+            var c = new Vertex(new Vector3(0, 0.5, 1.0));
+            var t = new Triangle(a, b, c);
+            var r = new Ray(Vector3.Origin, Vector3.ZAxis);
+            Vector3 xsect;
+            Assert.True(r.Intersects(t, out xsect));
+
+            r = new Ray(Vector3.Origin, Vector3.ZAxis.Negate());
+            Assert.False(r.Intersects(t, out xsect));
+        }
+
+        [Fact]
+        public void IntersectsAtVertex()
+        {
+            var a = new Vertex(new Vector3(-0.5, -0.5, 1.0));
+            var b = new Vertex(new Vector3(0.5, -0.5, 1.0));
+            var c = new Vertex(new Vector3(0, 0.5, 1.0));
+            var t = new Triangle(a, b, c);
+            var r = new Ray(new Vector3(-0.5, -0.5, 0.0), Vector3.ZAxis);
+            Vector3 xsect;
+            Assert.True(r.Intersects(t, out xsect));
+        }
+
+        [Fact]
+        public void IsParallelTo()
+        {
+            var a = new Vertex(new Vector3(-0.5, -0.5, 1.0));
+            var b = new Vertex(new Vector3(0.5, -0.5, 1.0));
+            var c = new Vertex(new Vector3(0, 0.5, 1.0));
+            var t = new Triangle(a, b, c);
+            var r = new Ray(Vector3.Origin, Vector3.XAxis);
+            Vector3 xsect;
+            Assert.False(r.Intersects(t, out xsect));
+        }
+
+        [Fact]
+        public void RayIntersectsTopography()
+        {
+            this.Name = "RayIntersectTopo";
+
+            var elevations = new double[25];
+
+            int e = 0;
+            for (var x = 0; x < 5; x++)
+            {
+                for (var y = 0; y < 5; y++)
+                {
+                    elevations[e] = Math.Sin(((double)x / 5.0) * Math.PI) * 10;
+                    e++;
+                }
+            }
+            var topo = new Topography(Vector3.Origin, 4, elevations);
+            this.Model.AddElement(topo);
+
+            var modelPoints = new ModelPoints(new List<Vector3>(), new Material("begin", Colors.Blue));
+            this.Model.AddElement(modelPoints);
+            foreach (var t in topo.Mesh.Triangles)
+            {
+                var c = Center(t);
+                var o = new Vector3(c.X, c.Y);
+                modelPoints.Locations.Add(o);
+
+                var ray = new Ray(o, Vector3.ZAxis);
+
+                Vector3 xsect;
+                if (ray.Intersects(t, out xsect))
+                {
+                    try
+                    {
+                        var l = new Line(o, xsect);
+                        var ml = new ModelCurve(l);
+                        this.Model.AddElement(ml);
+                    }
+                    catch
+                    {
+                        continue;
+                    }
+                }
+            }
+        }
+
+
+        [Fact]
+        public void RayIntersectsSolidOperation()
```

Comment:
Can we break up these assertions into separate tests like. I'll suggest some names in line below.

Reply:
There are also a couple of checks that we should add:
- Ray intersection at a vertex.
- Ray intersection when the ray origin is in the plane of a face.

Reply:
What should the expected behavior be in these cases? intersection or no?

Reply:
I note that IntersectsAtVertex assumes that intersects = true. But if a ray lies exactly along an edge, the intersection is a line, so could not be meaningfully reported without loss of information (ie only one or two vertices). Similarly, if the ray origin is in the plane of the face, the ray might also be parallel to that face and intersect it again, along infinite points. My inclination is actually to always *exclude* vertices/ edges / parallel along faces, but there may be a good reason to want these included and a sensible "result" to return as well.

Reply:
Good question. In the case of the vertex, I'd return the vertex. In the case of the ray origin in the plane of the face (an also inside the face's loop), I'd return the origin.

Reply:
Done.